src/include/miopen/groupnorm/problem_description.hpp

```diff
@@ -0,0 +1,162 @@
+
+#pragma once
+
+#include <miopen/problem_description_base.hpp>
+#include <miopen/activ.hpp>
+#include <miopen/tensor.hpp>
+
+#include <cassert>
+#include <string>
+
+namespace miopen {
+
+struct NetworkConfig;
+
+namespace groupnorm {
+
+struct ProblemDescription : ProblemDescriptionBase
+{
+    ProblemDescription(miopenNormMode_t mode_,
+                       const TensorDescriptor& xDesc_,
+                       const TensorDescriptor& weightDesc_,
+                       const TensorDescriptor& biasDesc_,
+                       const TensorDescriptor& yDesc_,
+                       const TensorDescriptor& meanDesc_,
+                       const TensorDescriptor& rstdDesc_,
+                       uint64_t num_groups_,
+                       float epsilon_)
+        : mode(mode_),
+          xDesc(xDesc_),
+          weightDesc(weightDesc_),
+          biasDesc(biasDesc_),
+          yDesc(yDesc_),
+          meanDesc(meanDesc_),
+          rstdDesc(rstdDesc_),
+          num_groups(num_groups_),
+          epsilon(epsilon_)
+    {
+        if(xDesc.GetType() != yDesc.GetType())
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "GroupNormForward: Tensor types do not match.");
+        }
+        if(meanDesc.GetType() != rstdDesc.GetType())
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "GroupNormForward: Tensor types do not match.");
+        }
+        if(xDesc.GetLengths() != yDesc.GetLengths())
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "GroupNormForward: Tensor dimension lengths do not match.");
+        }
+        if((num_groups < 1) || (xDesc.GetLengths().size() < 3) ||
+           (xDesc.GetLengths()[1] % num_groups != 0))
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "GroupNormForward: The channel size of input tensor should be divisible "
+                         "by num_groups.");
+        }
+        if(!(xDesc.IsPacked() && weightDesc.IsPacked() && biasDesc.IsPacked() && yDesc.IsPacked() &&
+             meanDesc.IsPacked() && rstdDesc.IsPacked()))
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "GroupNormForward: Unpacked tensors are not supported.");
+        }
+    }
+
+    miopenNormMode_t GetMode() const { return mode; }
+    const TensorDescriptor& GetXDesc() const { return xDesc; }
+    const TensorDescriptor& GetWeightDesc() const { return weightDesc; }
+    const TensorDescriptor& GetBiasDesc() const { return biasDesc; }
+    const TensorDescriptor& GetYDesc() const { return yDesc; }
+    const TensorDescriptor& GetMeanDesc() const { return meanDesc; }
+    const TensorDescriptor& GetRstdDesc() const { return rstdDesc; }
+    int32_t GetNumGroups() const { return num_groups; }
+    float GetEpsilon() const { return epsilon; }
+
+    bool IsSameType() const
+    {
+        if(xDesc.GetType() != yDesc.GetType())
+        {
+            return false;
+        }
```

Comment:
This is currently dead code because such an object cannot be created.

I would say that checks that cause construction failure should be removed from helper functions (and some helper functions could be removed entirely).

Reply:
I removed them.

Reply:
[resolved]